### PR TITLE
[jit] Fix handling of IntList[k] parameters

### LINKED ISSUE #6734: [JIT] jit does not properly handle IntList[x] params
For example, `adaptive_avg_pool1d` has an `IntList[1]` parameter as the second argument in `native_functions.yaml`.
```
@torch.jit.script
def func(x):
  return torch.nn.functional.adaptive_avg_pool1d(x, output_size=1)
```
gives
```
unknown builtin op:
@torch.jit.script
def func(x):
    return torch.nn.functional.adaptive_avg_pool1d(x, output_size=1)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
```
whereas 
```
@torch.jit.script
def func(x):
  return torch.nn.functional.adaptive_avg_pool1d(x, output_size=[1])
```
works. When you remove the `output_size=` it's the other way around.

This is on current master.
I came across this because it causes the build failure in #6152 .

In terms of work-around, maybe one could generate `i` and `is` signatures when one encounters `IntList[x]`.

Best regards

Thomas

### PR DESCRIPTION
**Pull request by elanmart**
Fixes https://github.com/pytorch/pytorch/issues/6734

If there's anything wrong with this PR, it's my fault. If there's anything good about it, credit goes to @t-vi.

### CODE REVIEW

**Comment by ezyang** [comment]
CC @apaszke @zdevito @jamesr66a

File: tools/jit/gen_jit_dispatch.py [hunk 1]
```diff
@@ -91,20 +99,24 @@ def gen_jit_dispatch(declarations, out):
     def is_tensor_arg(arg):
         return arg['simple_type'] in {'Tensor', 'TensorList'}
 
+    def is_sized_intlist_arg(arg):
```
Comment (apaszke) [inline tools/jit/gen_jit_dispatch.py]: Would be nice to add a comment because it's not obvious what a "sized intlist" is. Can you add a docstring saying sth like
```
"""Returns true for args declared as IntList[x], but not IntList."""
```

File: tools/jit/gen_jit_dispatch.py [hunk 2]
```diff
@@ -195,9 +220,18 @@ def emit_decl(decl):
         all_arguments_are_inputs = tuple(True for _ in arguments)
         only_tensors_are_inputs = tuple(is_tensor_arg(arg) for arg in arguments)
 
+        sized_intlist_args = [i for i, a in enumerate(arguments) if is_sized_intlist_arg(a)]
+        inds_to_replace_powerset = (set(c)
```
Comment (apaszke) [inline tools/jit/gen_jit_dispatch.py]: Is there any reason to treat them as a set? `combinations` are without replacement, so it should be ok to use tuples.

File: tools/jit/gen_jit_dispatch.py [hunk 3]
```diff
@@ -195,9 +220,18 @@ def emit_decl(decl):
...
+                                    for r in range(len(sized_intlist_args) + 1)
+                                    for c in combinations(sized_intlist_args, r))
+
         # NB: if there are no scalar args then both options on LHS are equivalent, so deduplicate them.
-        for variant in set([all_arguments_are_inputs, only_tensors_are_inputs]):
-            emit_decl_variant(decl, variant, has_tensorlist)
+        for variant in {all_arguments_are_inputs, only_tensors_are_inputs}:
```
Comment (apaszke) [inline tools/jit/gen_jit_dispatch.py]: Can you please add a comment saying that the set here is important, because those two options might overlap? It's not exactly part of this PR, but you've touched it, and I realized that it's not obvious

File: tools/jit/gen_jit_dispatch.py [hunk 4]
```diff
@@ -91,20 +99,24 @@ def gen_jit_dispatch(declarations, out):
...
+        return (arg['simple_type'] == 'IntList') and ('size' in arg)
+
     def get_invocation(decl, args):
         if 'namespace' in decl['method_of']:
             return CALL_NAMESPACE.substitute(name=decl['name'], args=args)
         else:
             return CALL_METHOD.substitute(name=decl['name'], first=args[0], args=args[1:])
 
-    def emit_decl_variant(decl, is_positional_arg, has_tensorlist):
+    def emit_decl_variant(decl, is_positional_arg, has_tensorlist, indices_to_replace=None):
```
Comment (apaszke) [inline tools/jit/gen_jit_dispatch.py]: As far as I can see this only affects the attribute case, but I don't think this is necessary. Instead of going through the powerset and blowing up the number of cases we could just alter the way we parse attribute values, similarly to what we do with positional args. This has both better complexity, and doesn't require this extra argument with pretty annoying semantics.

**Review by apaszke — COMMENTED** [review]
Thanks a lot @elanmart and @t-vi! I think we can do this without going through the powerset, so it would be nice to change it.

Actually I think it would be even better if we had a different auto-generated JIT pass, that normalizes out all of those cases that are normally handled by the Python arg parser. Right now it's hard to handle those nodes in all optimization passes, because they can appear in a lot of possible configurations. What we should do instead is do this normalization before even attempting to run anything else on the graph. In this case, it would check if a particular node has `IntList[x]` args declared, and would change any `int` attributes under this name to correctly expanded `IntList` attrs. However that's a slightly bigger change, so I'm ok with merging this to unblock the sum PR, and I'll fix it later myself.

**Comment by t-vi** [comment]
So I must admit that the faults in the implementation strategy are probably mine here.

So this is what I had in mind when I looked into making the JIT actually use python_arg_parser code for determining the function to use.
So when I think about this, I would look into changing this essentially to imitating (or using) `python arg parser.h/cpp` in that we lookup the name and then go through the iterated positions as the parameters processed by the JIT are not automatically converted into the right descriptor string.
However, that would be a largish change. In my preliminary patch, I implemented a FunctionSignature.parse_jit (or match_jit) function and a FunctionParameter.check_jit function to do this. Then the jit dispatch would have the full signature - it might also be useful to handle the "exotic" parameters such as device)
My understanding is that that is not the way you want, but what would be the way you would want the jit  to work here?

File: tools/jit/gen_jit_dispatch.py (hunk 3 above)
Comment (elanmart) [reply tools/jit/gen_jit_dispatch.py]: Hm? You mean the comment above? `# NB: if there are no scalar args then both options on LHS are equivalent, so deduplicate them.`

I didn't touch it

File: tools/jit/gen_jit_dispatch.py (hunk 2 above)
Comment (elanmart) [reply tools/jit/gen_jit_dispatch.py]: Nah, it was an impulse since later I test `if (value in collection)`. This ofc. doesn't matter for collections of length ~4.

File: test/test_jit.py [hunk 5]
```diff
@@ -2833,6 +2833,29 @@ def foo(x, y):
...
+        @torch.jit.script
+        def foo_3(x):
+            return torch.nn.functional.adaptive_avg_pool1d(x, output_size=[1])
+
+        x = torch.randn(8, 8, 8)
+        r = foo_0(x)
+
+        self.assertEqual(r, foo_1(x))
+        self.assertEqual(r, foo_2(x))
+        self.assertEqual(r, foo_3(x))
```
[the remark this answers is not in the document]
Comment (elanmart) [reply test/test_jit.py]: Sure, sorry, didn't know about it.

**Comment by elanmart** [comment]
@apaszke 

> Actually I think it would be even better if we had a different auto-generated JIT pass, that normalizes out all of those cases that are normally handled by the Python arg parser. (...) However that's a slightly bigger change, so I'm ok with merging this to unblock the sum PR, and I'll fix it later myself.

I agree this PR doesn't seem like the nicest possible change. I'd be happy to give a try at a more elaborate one, but I'm afraid wouldn't know where to start with this.

So perhaps we can unblock #6152 , and I'll try to work with @t-vi to come up with a proper solution?

**Comment by elanmart** [comment]
```bash
$ wc -l torch/csrc/jit/generated/aten_dispatch.cpp 
21619 torch/csrc/jit/generated/aten_dispatch.cpp
```

**Comment by apaszke** [comment]
Yeah, so the powerset makes it 2x larger 😕
```
wc -l torch/csrc/jit/generated/aten_dispatch.cpp 
10897 torch/csrc/jit/generated/aten_dispatch.cpp
```
Eh, it's not great, but I guess we'll need to stick with it for now.

File: tools/jit/gen_jit_dispatch.py (hunk 4 above)
Comment (apaszke) [reply tools/jit/gen_jit_dispatch.py]: Can you please still change this one? That's the only blocker for this PR.

File: tools/jit/gen_jit_dispatch.py (hunk 3 above)
Comment (apaszke) [reply tools/jit/gen_jit_dispatch.py]: Nvm, I'll do that in a followup PR.

**Review by apaszke — DISMISSED** [review]
Let's clean up `indices_to_replace` and merge this.

File: tools/jit/gen_jit_dispatch.py (hunk 4 above)
Comment (elanmart) [reply tools/jit/gen_jit_dispatch.py]: I'm really sorry, could you clarify that a bit?

You'd like to move the whole powerset-generating logic [here](https://github.com/elanmart/pytorch/blob/d7103921ca3d3945636e128b2e6884fde21ded98/tools/jit/gen_jit_dispatch.py#L155)?

Comment (apaszke) [reply tools/jit/gen_jit_dispatch.py]: Oh sorry, my bad. I tought I left a different comment here, and didn't bother to read it again 😕 A better solution would be to take a dict that's a mapping from either arg indices or arg names to their new types (and call it e.g. `override_types`). This is similar in implementation, but has much clearer semantics.

**Review by ezyang — APPROVED** [review]
